### PR TITLE
build: Adding OFI support for MPI

### PR DESCRIPTION
**Pull request by Nikhil-Nunna**
## Description
This PR adds OFI support for MPI which will enable
performant execution on systems that leverage OFI rather than UCX. 
The PR also updates version of PyTorch for the tf2-cpu
target from 1.10 to 1.11. This was done because after building the
tf2-cpu image and converting it to Singularity, a simple PyTorch test
failed with an error similar to:
```
Singularity> python -c "import torch; print(torch.__version__)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/conda/lib/python3.8/site-packages/torch/__init__.py",
  line 29, in <module>
    from .torch_version import __version__ as __version__
  File
  "/opt/conda/lib/python3.8/site-packages/torch/torch_version.py",
  line 3, in <module>
    from pkg_resources import packaging  # type: ignore[attr-defined]
  File
  "/opt/conda/lib/python3.8/site-packages/pkg_resources/__init__.py",
  line 3267, in <module>
    def _initialize_master_working_set():
<...snip...>
  line 2780, in _get_metadata
    for line in self.get_metadata_lines(name):
  File
  "/opt/conda/lib/python3.8/site-packages/pkg_resources/__init__.py",
  line 1911, in get_metadata_lines
    return yield_lines(self.get_metadata(name))
  File
  "/opt/conda/lib/python3.8/site-packages/pkg_resources/__init__.py",
  line 1898, in get_metadata
    with io.open(self.path, encoding='utf-8', errors="replace") as f:
PermissionError: [Errno 13] Permission denied: '/opt/conda/lib/python3.8/site-packages/.wh.setuptools-58.0.4-py3.8.egg-info'
```
It seemed that the install of a different package in the image
installed a newer setuptools, which caused the conflict. Updating
PyTorch to 1.11 fixed this, as shown below:
```
Singularity> python -c "import torch; print(torch.__version__)"
1.11.0+cpu
```
Tested the tf2-cpu target build and did simple tests
with PyTorch and TF with each to verify it works.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -90,7 +100,9 @@ build-cpu-py-38-base:
 	    --platform "$(PLATFORMS)" \
 		--build-arg BASE_IMAGE="$(UBUNTU_IMAGE_TAG)" \
 		--build-arg PYTHON_VERSION="$(PYTHON_VERSION)" \
+		--build-arg UBUNTU_VERSION="ubuntu20.04" \
```
Comment (mpkouznetsov) [inline Makefile:103]: use UBUNTU_VERSION constant defined on line 23 in this file

File: Makefile [hunk 2]
```diff
@@ -77,7 +85,9 @@ build-cpu-py-37-base:
 	docker build -f Dockerfile-base-cpu \
 		--build-arg BASE_IMAGE="$(UBUNTU_IMAGE_TAG)" \
 		--build-arg PYTHON_VERSION="$(PYTHON_VERSION_37)" \
-		--build-arg "$(MPI_BUILD_ARG)" \
+		--build-arg UBUNTU_VERSION="ubuntu20.04" \
```
Comment (mpkouznetsov) [inline Makefile:88]: use UBUNTU_VERSION constant defined on line 23 in this file

File: Makefile [hunk 3]
```diff
@@ -77,7 +85,9 @@ build-cpu-py-37-base:
...
+        --build-arg "$(MPI_BUILD_ARG)" \
+		--build-arg "$(OFI_BUILD_ARG)" \
```
Comment (mpkouznetsov) [inline Makefile:90]: I think this resolves to an empty string, for example, here: https://app.circleci.com/pipelines/github/determined-ai/environments/363/workflows/27d11f93-746e-44a2-8720-6f604d882e77/jobs/2798

```
#!/bin/bash -eo pipefail
make build-tf1-cpu NGC_PUBLISH= REGISTRY_REPO=environments-dev WITH_MPI=1
docker build -f Dockerfile-base-cpu \
	--build-arg BASE_IMAGE="ubuntu:20.04" \
	--build-arg PYTHON_VERSION="3.7.11" \
	--build-arg UBUNTU_VERSION="ubuntu20.04" \
        --build-arg "USE_GLOO=1" \
	--build-arg "" \
	-t determinedai/environments-dev:py-3.7-base-cpu-6ade8bf \
	-t determinedai/environments-dev:py-3.7-base-cpu-0.19.1 \
	.
invalid argument "" for "--build-arg" flag: invalid environment variable: 
See 'docker build --help'.
make: *** [Makefile:85: build-cpu-py-37-base] Error 125

Exited with code exit status 2
CircleCI received exit code 2
```

**Review by mpkouznetsov — COMMENTED** [review]
A couple of suggestions. Also, the build with your changes fails so definitely some things to fix.

File: Makefile [hunk 4]
```diff
@@ -77,7 +94,9 @@ build-cpu-py-37-base:
 	docker build -f Dockerfile-base-cpu \
 		--build-arg BASE_IMAGE="$(UBUNTU_IMAGE_TAG)" \
 		--build-arg PYTHON_VERSION="$(PYTHON_VERSION_37)" \
-		--build-arg "$(MPI_BUILD_ARG)" \
+		--build-arg UBUNTU_VERSION="$(UBUNTU_VERSION)" \
+        --build-arg "$(MPI_BUILD_ARG)" \
```
Comment (mpkouznetsov) [inline Makefile:98]: the indentation looks odd here.

File: Makefile [hunk 5]
```diff
@@ -25,25 +25,42 @@ UBUNTU_IMAGE_TAG := ubuntu:20.04
...
+	OFI_BUILD_ARG := WITH_OFI
+	HOROVOD_WITH_MPI := 0
+	HOROVOD_WITHOUT_MPI := 1
+	HOROVOD_CPU_OPERATIONS := GLOO
+	MPI_BUILD_ARG := USE_GLOO=1
 endif
 DEEPSPEED_VERSION := 0.7.0
 
+ifneq (,$(findstring linux/arm64,$(PLATFORMS)))
+	TORCH_PIP_VERSION_TF2_Name := "1.10"
```
Comment (mpkouznetsov) [inline Makefile:57]: odd indentation

File: Makefile [hunk 6]
```diff
@@ -25,25 +25,42 @@ UBUNTU_IMAGE_TAG := ubuntu:20.04
...
+	WITH_MPI := 0
+	OFI_BUILD_ARG := WITH_OFI
+	HOROVOD_WITH_MPI := 0
+	HOROVOD_WITHOUT_MPI := 1
+	HOROVOD_CPU_OPERATIONS := GLOO
+	MPI_BUILD_ARG := USE_GLOO=1
 endif
 DEEPSPEED_VERSION := 0.7.0
 
+ifneq (,$(findstring linux/arm64,$(PLATFORMS)))
```
Comment (mpkouznetsov) [inline Makefile:56]: Am I missing something or can the two branches of this conditional statement be moved into the previous conditional statement? I mean, `PLATFORMS` contains linux/arm64 if and only if `$(WITH_MPI)` is not equal "1".

File: dockerfile_scripts/ompi.sh [hunk 7]
```diff
@@ -2,11 +2,13 @@
 
 set -x
 
-# Install the Mellanox OFED stack.  Note that this is dependent on what 
+# Install the Mellanox OFED stack.  Note that this is dependent on what
 # the base OS is (ie, Ubuntu 20.04) so if that changes then this needs updated.
 #MOFED_VER=5.0-2.1.8.0
-MOFED_VER=5.5-1.0.3.2
+#MOFED_VER=5.5-1.0.3.2
+MOFED_VER=5.4-3.4.0.0
```
Comment (cdrickett) [inline dockerfile_scripts/ompi.sh:9]: Is there a reason you’re using 5.4 instead of 5.5?

**Review by cdrickett — APPROVED** [review]

File: dockerfile_scripts/ompi.sh (hunk 7 above)
Comment (Nikhil-Nunna) [reply dockerfile_scripts/ompi.sh:9]: I tried to wget 5.5 and it couldn't find it so I tried 5.4 and that worked.

**Review by mpkouznetsov — APPROVED** [review]

**Pull request merged**